File: tier3_advanced/modified_binary_search/problems/p410_split_array_sum.c
/*
 * P410: Split Array Largest Sum (Hard)
 * https://leetcode.com/problems/split-array-largest-sum/
 * Topics: Array, Binary Search, Dynamic Programming, Greedy, Prefix Sum
 * 
 * Given an integer array nums and an integer k, split nums into k non-empty subarrays such that the largest sum of any subarray is minimized.
 * Return the minimized largest sum of the split.
 * A subarray is a contiguous part of the array.
 * 
 * Example 1:
 *     Input: nums = [7,2,5,10,8], k = 2
 *     Output: 18
 *     Explanation: There are four ways to split nums into two subarrays.
 *     The best way is to split it into [7,2,5] and [10,8], where the largest sum among the two subarrays is only 18.
 * 
 * Example 2:
 *     Input: nums = [1,2,3,4,5], k = 2
 *     Output: 9
 *     Explanation: There are four ways to split nums into two subarrays.
 *     The best way is to split it into [1,2,3] and [4,5], where the largest sum among the two subarrays is only 9.
 * 
 * Constraints:
 *     - 1 <= nums.length <= 1000
 *     - 0 <= nums[i] <= 106
 *     - 1 <= k <= min(50, nums.length)
 * 
 * Template (python3):
 *     class Solution:
 *         def splitArray(self, nums: List[int], k: int) -> int:
 * 
 * Hint: Binary search the answer in [max(nums), sum(nums)] and check feasibility greedily.
 */
#include <stdio.h>
#include <limits.h>

typedef struct {
    const char *label;
    int nums[1000];
    int nums_n;
    int k;
    long long expected;
} TestCase;

static int feasible(int *nums, int nums_n, long long max_sum, int k)
{
    abort();
}

long long splitArray(int *nums, int numsSize, int k)
{
    abort();
}

int main(void)
{
    TestCase tests[] = {
        {"example 1", {7,2,5,10,8}, 5, 2, 18},
        {"example 2", {1,2,3,4,5}, 5, 2, 9},
        {"example 3", {1,4,4}, 3, 3, 4},
        {"k equals length", {1,2,3,4,5}, 5, 5, 5},
        {"single element", {5}, 1, 1, 5},
        {"all zeros", {0,0,0,0}, 4, 2, 0},
        {"uniform values", {1,1,1,1,1,1,1,1}, 8, 4, 2},
        {"large values", {1000000,1000000,1000000}, 3, 2, 2000000},
    };
    int n = sizeof(tests) / sizeof(tests[0]);

    printf("\n============================================================\n");
    printf("  410. Split Array Largest Sum\n");
    printf("============================================================\n");
    int passed = 0;
    for (int i = 0; i < n; i++) {
        long long got = splitArray(tests[i].nums, tests[i].nums_n, tests[i].k);
        if (got == tests[i].expected) {
            passed++;
            printf("  Test %d (%s): PASS\n", i + 1, tests[i].label);
        } else {
            printf("  Test %d (%s): FAIL\n", i + 1, tests[i].label);
            printf("    Expected: %lld\n    Got:      %lld\n", tests[i].expected, got);
        }
    }
    printf("\n  %d/%d passed\n", passed, n);
    printf("============================================================\n\n");
    return passed == n ? 0 : 1;
}

File: tier3_advanced/modified_binary_search/test_runners/p410_split_array_sum.py
from src.utils.judge_base import JudgeBase, TestCase, register


class P410SplitArraySum(JudgeBase):
    name = "410. Split Array Largest Sum"
    test_cases = [
        TestCase(input=([7, 2, 5, 10, 8], 2), expected=18, label="example 1"),
        TestCase(input=([1, 2, 3, 4, 5], 2), expected=9, label="example 2"),
        TestCase(input=([1, 4, 4], 3), expected=4, label="example 3"),
        TestCase(input=([1, 2, 3, 4, 5], 5), expected=5, label="k equals length"),
        TestCase(input=([5], 1), expected=5, label="single element"),
        TestCase(input=([0, 0, 0, 0], 2), expected=0, label="all zeros"),
        TestCase(input=([1, 1, 1, 1, 1, 1, 1, 1], 4), expected=2, label="uniform values"),
        TestCase(input=([1000000, 1000000, 1000000], 2), expected=2000000, label="large values"),
    ]

    def to_stdin(self, tc_input) -> str:
        return self.serialize_multi(
            self.serialize_int_array(tc_input[0]),
            self.serialize_int(tc_input[1]),
        )

    def check_stdout(self, stdout: str, tc_expected) -> bool:
        return self.parse_int(stdout) == tc_expected


register(P410SplitArraySum)
